### ISSUE TITLE
[BUG] Incorrect batch_size used for ThroughputTimer

### ISSUE THREAD

**Issue by clumsy** [microsoft/DeepSpeed#2498]
**Describe the bug**
`ThroughputTimer` is started/stopped for model steps, but in `runtime/engine` it uses the micro_batch_size only:
[here](https://github.com/microsoft/DeepSpeed/blob/ee39187d8f07f7efc615f64affa8cc60ccf41eb5/deepspeed/runtime/engine.py#L297)

Compare the above line to the following that uses the right value in `runtime/pipe/engine`: [here](https://github.com/microsoft/DeepSpeed/blob/b8fb9c3f1a8a8e3e574a7d53e83ce2b72d471aa3/deepspeed/runtime/pipe/engine.py#L107)

**To Reproduce**
1. Enable `wall_clock_breakdown`

**Expected behavior**
This timer should use  `self.train_micro_batch_size_per_gpu() * self.gradient_accumulation_steps()`.

**ds_report output**
N/A

**Screenshots**
N/A

**System info (please complete the following information):**
 - OS: `Amazon Linux 2`
 - GPU count and types: `8 x A100`
 - Interconnects (if applicable): `4 x 100 Gbps`
 - Python version: `3.8`
 - Any other relevant info about your setup: 

**Launcher context**
Launching the experiment with the `deepspeed` launcher.

**Docker context**
Can't share

**Additional context**
DeepSpeed v0.7.3

**Comment by clumsy** [comment]
Hi @ShijieZZZZ, can you please check my PR? There's a weird environment issue for one of the checks, while others run fine.

**Comment by ShijieZZZZ** [comment]
Hi @clumsy, checked tests are passed. 
Are you using [avg_sample_per_sec()](https://github.com/microsoft/DeepSpeed/blob/ee39187d8f07f7efc615f64affa8cc60ccf41eb5/deepspeed/utils/timer.py#L218) to compute throughput?

**Comment by clumsy** [comment]
On the second thought, I'm a bit confused now.

The current throughput number _could_ be consistent if the steps it refers to are really __micro steps__.

In the docs [here](https://www.deepspeed.ai/docs/config-json/) `steps_per_print` are referred to as "training steps". So my assumption is that it should be printed at gradient accumulation boundaries, after full step processing `train_batch_size` number samples.

If it's intended to be printed at micro steps I wonder how useful it is when the value fluctuates between intermediate micro steps and micro step at the gradient boundary where additional operations are executed slowing down the throughput.

I'll adjust my PR to report for full steps just in case this is how it was supposed to be.

**Comment by clumsy** [comment]
So is my understanding correct, @ShijieZZZZ? Is the throughput timer supposed to report per full step, not micro step as it does currently? If yes, please check out the recent version of my PR? The unit tests get 5xx from HuggingFace though.

**Comment by ShijieZZZZ** [comment]
Hi @clumsy, I agree that throughput time should report every gradient_accumulation_steps, I will take a look your PR.

**Comment by ShijieZZZZ** [comment]
@clumsy, please check this. https://github.com/microsoft/DeepSpeed/pull/2553

**Comment by tjruwase** [comment]
Closing. Please re-open if needed.

**Comment by clumsy** [comment]
Sorry @tjruwase and @ShijieZZZZ, I was just checking the suggested fix and I believe the problem is still there, in part. While I believe the timer should now report throughput at model steps, `curr_samples_sec` is still not accurate. It is still effectively `(self.train_micro_batch_size_per_gpu() * self.num_workers) / duration` (does not take gradient accumulation (micro) steps into account [here](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/engine.py#L300)) which was the main goal of this issue.

**Comment by clumsy** [comment]
Also we seem to ignore `steps_per_output` after the latest change. I found one problem with my original PR but I'll fix it and keep the tests.
I think the root of all issues here is the fact that `steps` inside the timer are really micro steps. So if we pass gradient accumulation number we can use it to both multiply micro batch size per GPU and mod the number of start/stop invocations to detect model steps. And once `steps_per_output` is a factor of the number of micro steps divided by gradient accumulation number - we can report the value.

**Comment by tjruwase** [comment]
@clumsy, thanks for the analysis. I agree that there are still issues to resolve. However, one issue that I notice with your proposal is that it will introduce the concept of gradient accumulation into the timer, which I think it unnecessary and makes the timer less generic. I would like a design where the timer remains simple and the client (i.e., the engine) incurs the complexity of using the timer to meet its needs. Also, for simplicity timer `start()` and `stop()` are called on micro steps, and so it is better for timer to be initialized with `micro_batch_size` so that the state is always consistent, e.g., for debugging. 

So, I am thinking of an alternative approach:

1. Rename timer fields for clarity, `local_step_count` as `micro_step_count`, and `total_step_count` as `global_step_count`.
2. Redefine timer stop as `def stop(self, global_step=False, report_speed=False)`
3. On gradient accumulation boundaries, client calls `timer.stop()` with `global_step=True`, and `report_speed=True` if rank 0, otherwise call both flags are False
4. timer stop() updates `micro_step_count` always, updates `global_step_count` if `global_step=True` and prints if `report_speed=True`

What do you think?

**Comment by clumsy** [comment]
Right, @tjruwase, this is exactly what I was thinking as well. In fact, I already prepared a PR which mostly follows your proposal. Please let me know if you want me to adjust it.

**Comment by tjruwase** [comment]
@clumsy, thanks for sharing the PR. I already added some comments. Please let me know your thoughts.

**Comment by clumsy** [comment]
> @clumsy, thanks for sharing the PR. I already added some comments. Please let me know your thoughts.

Thank you so much, pushed another revision. Please let me know if it looks better now.
